### ISSUE TITLE
ConFIG applicable to VAEs?

### ISSUE THREAD

**Issue by juliusaka** [tum-pbs/ConFIG#7]
Dear ConFIG Team,
Thank you for your great work and the well documented code.
In my work, I am working with VAEs and an extension of VAE to NeuralODE and thereby focus on dimensionality reduction. So for that I typically have the situation that my loss is

`loss = reconstruction_loss + beta * kl_divergence`

and kl_divergence facilitates the dimensionality reduction by enforcing the parameterized distribution in latent space to be of mu=0 and sigma=1. 
I wondered if ConFIG helps to improve training of VAE by scaling the gradients in a way that both goals (reconstruction and dimensionality reduction) are reached. 
However, (at least in my implementation) it did not work. I'm using a standard VAE trained together with a regressor model that tracks the position of labels  in the latent space. For that, the loss is

`loss = reconstruction_loss + beta * kl_divergence + regressor_loss`

I then thought to first start with Adam and without ConFIG and then switch to using ConFIG. Something interesting happend:

![Image](https://github.com/user-attachments/assets/dadc293a-c9be-4608-b65b-9c99fbabc3b4)
x-axis: time in seconds, y-axis in log-scale

After switching in epoch 200 to using ConFIG, MSE_loss detoriates, while the KL_loss is reduced to 0, the overall loss gets very high again. I used a beta of 0.0001. 

I wonder if you have an explanation why ConFIG does not work for this. Might the cause for this be that reconstruction_loss and kl_divergence both target the same model parameters for the same input?

I also append my implementation in the training routine; I hope that there is no mistake in it. 

```python
grads = []
for i, loss_i in enumerate(ret_vals_train['loss_list']):
    model.zero_grad(set_to_none=True)
    loss_i.backward(retain_graph=True )
    grads.append(get_gradient_vector(model, none_grad_mode='zero'))
grad_config = ConFIG_update(grads, losses = ret_vals_train['loss_list'])
optimizer.zero_grad(set_to_none=True)
apply_gradient_vector(model,grad_config, none_grad_mode='zero', zero_grad_mode='pad_value')
optimizer.step()
```
(I return a list of loss components from my model evaluation. In your examples your evaluated the model for each loss component again, I just calculated the backward pass repeatingly again while setting grads after each backward pass to 0)

I'm looking forward to your take on this,
Julius

**Comment by qiauil** [comment]
Hi, @juliusaka 

Sorry for the late reply. You can directly mention me, @qiauil, next time you have any issues with the code since the notification from the group member doesn't work well...

From my understanding, your issue is due to the Adam optimizer. Adam optimizer uses the expansional moving average of the gradients to do the optimization steps. Thus, once you shift from Adam to Adam+ConFIG, you are still not directly optimized with the config gradient but the ema of the gradients history plus a slight update provided by the ConFIG gradient. That will cause a mess. I would recommend directly starting with ConFIG at the beginning.

Note that even if you start with Adam, you should also be careful. As we discussed before, Adam uses the ema to do the optimized update. Even if the gradient provided by the ConFIG is conflict-free in each step, the ema of the gradients are not guaranteed to be conflict-free. Nonetheless, Adam will still perform well with ConFIG in most cases. I would also recommend https://tum-pbs.github.io/ConFIG/start/troubleshooting/ where we also discussed how the magnitude of different loss terms might effect the performance with Adam optimizer.

Cheers,
Qiang

**Comment by juliusaka** [comment]
Hey @qiauil,

I actually tested if ConFIG works starting with it from the beginning of the training, it does not work either. Then, to adress your concerns due to the Adam Optimizer, I switched to SGD, but unfortuanetly it did not work either. After your comment about this, I experimented with the WeightModel and LengthModel Feature and this seems to help enabling convergence (for both SGD and Adam). Thank you for that hint!
But I wonder if ConFIG overall is applicable to VAE. VAE (with the two loss components) optimize for a "compromise" between reconstruction and information bottleneck, while (as I understand your paper), ConFIG targeted PINNs to optimize for different aspects. Do you have any take on this?

Thanks a lot for your answer,
Julius

**Comment by qiauil** [comment]
Hi, @juliusaka 

Good to know that the length and weight model helps to stabilize the training.

The ConFIG method isn't specially optimized for the PINNs, so theoretically, I would agree that it is applicable to VAE.

But I would also like to mention one point. For any kind of multi-task optimization problem, including our ConFIG, the final target is to make the optimization converge to the Pareto front, as shown by a simple case at https://tum-pbs.github.io/ConFIG/examples/mtl_toy/. Let's take your example `reconstruction_loss + beta * kl_divergence.` The Pareto front refers to the state where you can't decrease the reconstruction loss without increasing the kl divergence. There are multiple points in the Pareto front where sometimes the reconstruction loss is higher and other times the kl divergence is higher. For the application where you really need to make the reconstruction loss (or the kl divergence) as small as possible, you may need to search again at the Pareto front. But in general, I think VAE doesn't require a special favor for one of the loss terms, and I think it would also make sense to use ConFIG there.

Cheers,
Qiang
